src/render.rs

```diff
@@ -0,0 +1,236 @@
+            InvitationRevoked => "had the invitation revoked",
+            ProfileChanged => "changed the display name or avatar",
+            NotImplemented => "performed an unimplemented operation",
+        };
+
+        let (prefix, color_action) = match change_op {
+            Joined | Invited => ("join", "green"),
+            _ => ("quit", "red"),
+        };
+
+        format!(
+            "{prefix}{color_user}{name} {color_deli}({color_reset}{state}{color_deli}){color_reset} {color_action}{op}",
+            prefix = Weechat::prefix(prefix),
+            color_user = Weechat::color("reset"), // TODO
+            color_deli = Weechat::color("chat_delimiters"),
+            color_action = Weechat::color(color_action),
+            color_reset = Weechat::color("reset"),
+            name = displayname,
+            state = self.state_key,
+            op = operation,
+        )
+    }
+
+    fn sender(&self) -> &UserId {
+        &self.sender
+    }
+}
+
+impl RenderableEvent for MessageEvent {
+    fn render(&self, displayname: &str) -> String {
+        use MessageEventContent::*;
+
+        let color_user = Weechat::color("green"); // TODO: get per-user color
+        let color_reset = Weechat::color("reset");
+        match &self.content {
+            // TODO: formatting for inline markdown and so on
+            Text(t) => format!(
+                "{color_user}{}{color_reset}\t{}",
+                displayname,
+                t.resolve_body(),
+                color_user = color_user,
+                color_reset = color_reset
+            ),
+            Emote(e) => format!(
+                "{color_user}{}{color_reset}\t{}",
+                displayname,
+                e.resolve_body(),
+                color_user = color_user,
+                color_reset = color_reset
+            ),
+            Audio(a) => format!(
+                "{color_user}{}{color_reset}\t{}: {}",
+                displayname,
+                a.body,
+                a.resolve_url(),
+                color_user = color_user,
+                color_reset = color_reset
+            ),
+            File(f) => format!(
+                "{color_user}{}{color_reset}\t{}: {}",
+                displayname,
+                f.body,
+                f.resolve_url(),
+                color_user = color_user,
+                color_reset = color_reset
+            ),
+            Image(i) => format!(
+                "{color_user}{}{color_reset}\t{}: {}",
+                displayname,
+                i.body,
+                i.resolve_url(),
+                color_user = color_user,
+                color_reset = color_reset
+            ),
+            Location(l) => format!(
+                "{color_user}{}{color_reset}\t{}: {}",
+                displayname,
+                l.body,
+                l.geo_uri,
+                color_user = color_user,
+                color_reset = color_reset
+            ),
+            Notice(n) => format!(
```

Comment:
Notices should be rendered as IRC style notices. Like [here](https://github.com/poljar/weechat-matrix/blob/master/matrix/buffer.py#L601) or in the IRC plugin.